### PR TITLE
[BEAM-4420] Add KafkaIO Integration Tests

### PR DESCRIPTION
**Pull request by lgajowy**
This pr adds basic IOIT that can be run for 1000 records (there's only hashcode for 1000 records currently). To make the code review easier and because I'm not a Kafka expert, I took a baby steps approach here - after this PR is approved consequent PRs will provide:

 - metrics collection
 - hashes for bigger loads 
 - jenkins job to run automatically and on-demand & dashboards for them

I ran this test on existing kubernetes kafka cluster that is there in .test-infra folder (with success). 

@aromanenko-dev @iemejia could you take a look?

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by iemejia** [comment]
Checkstyle not happy, you should use the vendored version of guava. Can you please also squash the commits into one with the proper title. Thanks.

File: sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOIT.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+      readResult.cancel();
+    }
+  }
+
+  private KafkaIO.Write<byte[], byte[]> writeToKafka() {
+    return KafkaIO.<byte[], byte[]>write()
+        .withBootstrapServers(options.getKafkaBootstrapServerAddress())
+        .withTopic(options.getKafkaTopic())
+        .withProducerConfigUpdates(
+            ImmutableMap.of("retention.ms", Integer.MAX_VALUE, "log.retention.bytes", -1))
```
[the remark this answers is not in the document]
Comment (lgajowy) [reply sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOIT.java]: According to the kafka docs you're right. My goal here was to ensure that every topic has the right retention policies before running the test. However, it looks that the kubernetes cluster already provides appropriate config for our tests: https://github.com/apache/beam/blob/0532e2a0e9544a63eca2cb38740f04344c1cea03/.test-infra/kubernetes/kafka-cluster/05-kafka/10broker-config.yml#L164

I'm removing the config updates here. Let me know if you're fine with the retention policy being provided in the cluster config and not in the test.

File: sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOIT.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        .withKeySerializer(ByteArraySerializer.class)
+        .withValueSerializer(ByteArraySerializer.class);
+  }
+
+  private KafkaIO.Read<byte[], byte[]> readFromKafka() {
+    return KafkaIO.readBytes()
+        .withBootstrapServers(options.getKafkaBootstrapServerAddress())
+        .withConsumerConfigUpdates(ImmutableMap.of("auto.offset.reset", "earliest"))
+        .withTopics(Lists.newArrayList(options.getKafkaTopic()))
```
[the remark this answers is not in the document]
Comment (lgajowy) [reply sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOIT.java]: True, thanks.

File: sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOIT.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+ * <p>NOTE: This test sets retention policy of the messages so that all messages are retained in the
+ * topic so that we could read them back after writing.
+ */
+@RunWith(JUnit4.class)
+public class KafkaIOIT {
+
+  /**
+   * Map of hash for 1000 uniformly distributed records with 10B keys and 90B values (100kB total).
+   */
+  private static final String EXPECTED_HASHCODE = "4507649971ee7c51abbb446e65a5c660";
```
[the remark this answers is not in the document]
Comment (lgajowy) [reply sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOIT.java]: Yes. However, it didn't make sense to wrap it to the map now for one element. I fixed the javadoc though.

File: sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOIT.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+    @Description("Kafka topic")
+    @Validation.Required
+    String getKafkaTopic();
+
+    void setKafkaTopic(String topic);
+
+    @Description("Filename prefix for file with hashcode")
+    @Validation.Required
+    String getHashcodeFilenamePrefix();
```
[the remark this answers is not in the document]
Comment (lgajowy) [reply sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOIT.java]: I forgot to remove that - thanks.

**Comment by lgajowy** [comment]
@iemejia @aromanenko-dev thanks for the review. I provided fixes. 

@iemejia I didn't squash the commits on purpose -  those are two atomic, independent changes. Having them as separate commits shows better why exactly something changed with proper commit titles and descriptions (see the description for the Kubernetes change). Squash will obfuscate this and will result with multiple unrelated changes in one commit. Other than that it will be easier to revert parts if needed. Suppose I was to revert only the Kubernetes change. Should I be forced to revert the whole test too?

File: sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOIT.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (aromanenko-dev) [reply sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOIT.java]: Yes, it's ok for me since it's used only for testing purposes.

**Comment by lgajowy** [comment]
Thanks!

**Pull request merged**